src/core/topology.rs

```diff
@@ -20,17 +20,56 @@
 
 //! Mixnet topology interface.
 
-use crate::{core::MixPublicKey, MixPeerId};
-
-/// Provide network topology information to the mixnet.
-pub trait Topology: Send + 'static {
-	/// Select a random recipient for the message to be delivered. This is
-	/// called when the user sends the message with no recipient specified.
-	/// E.g. this can select a random validator that can accept the blockchain
-	/// transaction into the block.
-	/// Return `None` if no such selection is possible.
-	fn random_recipient(&self) -> Option<MixPeerId>;
-
-	/// For a given peer return a list of peers it is supposed to be connected to.
-	fn neighbors(&self, id: &MixPeerId) -> Vec<(MixPeerId, MixPublicKey)>;
+use crate::{core::MixPublicKey, MixPeerId, core::MixPeerAddress};
+use rand::{prelude::{SliceRandom, IteratorRandom}, CryptoRng, Rng};
+
+/// Contains network information. Current implementation assumes that each node
+/// in the set is connected to every other node.
+#[derive(Default)]
+pub struct SessionTopology {
+	nodes: Vec<(MixPeerId, MixPublicKey, Vec<MixPeerAddress>)>,
+}
+
+impl SessionTopology {
+	pub fn new(nodes: Vec<(MixPeerId, MixPublicKey, Vec<MixPeerAddress>)>) -> Self {
+		Self { nodes }
+	}
+
+	pub fn random_recipient<R: Rng + CryptoRng + ?Sized>(&self, rng: &mut R) -> Option<MixPeerId> {
+		self.nodes.iter().choose(rng).map(|(id, _, _)| id.clone())
+	}
+
+	pub fn gateways<R: Rng + CryptoRng + ?Sized>(&self, rng: &mut R) -> Vec<MixPeerAddress> {
+		self.nodes
+			.as_slice()
+			.choose_multiple(rng, 5)
+			.map(|(_, _, addrs)| addrs.clone())
+			.flatten()
+			.collect()
+	}
+
+	pub fn random_path_to<R: Rng + CryptoRng + ?Sized>(
+		&self,
+		rng: &mut R,
+		start: &MixPeerId,
+		recipient: &MixPeerId,
+		num_hops: usize,
+	) -> Option<Vec<(MixPeerId, MixPublicKey)>> {
+		let (_, start_pk, _) = self.nodes.iter().find(|(id, _, _)| id == start)?;
+		let (_, last_pk, _) = self.nodes.iter().find(|(id, _, _)| id == recipient)?;
+
+		let mut result = vec![];
+		let mut prev = *start;
+		let mut next = *start;
+		let mut next_pk = *start_pk;
+		for i in 0 .. num_hops - 1 {
```

Comment:
Question, should we avoid using a same node twice (like in https://github.com/cheme/mixnet/blob/8e9c3046a8096b4896ffdff218d7b95b94ce0593/src/traits/set.rs#L107)?

Reply:
This code avoids duplicate consecutive hops, so that no node on the paths forwards the packet to itself (btw the forwading code currently does not support this). Other than that, it is fine for the node to be in included in the path multiple times.